### ISSUE TITLE
Exemption certificate validation fails during "Rules Validation"

### ISSUE THREAD

**Issue by mircoianese** [DevTrevi/DgcReader#84]
Hello, 

I am testing an Exemption Certificate that gets correctly validated by VerificaC19.

This is my code:

```
var _trustListProvider = ItalianTrustListProvider.Create(_httpClient, new ItalianTrustListProviderOptions()
            {
                RefreshInterval = TimeSpan.FromMinutes(refreshIntervalMinutes),
                MinRefreshInterval = TimeSpan.FromMinutes(minRefreshIntervalMinutes),
                MaxFileAge = TimeSpan.FromDays(60),
                BasePath = basePath,
                SaveCertificate = true,
                UseAvailableListWhileRefreshing = true
            });

var _rulesValidator = DgcItalianRulesValidator.Create(_httpClient, new DgcItalianRulesValidatorOptions
            {
                RefreshInterval = TimeSpan.FromMinutes(refreshIntervalMinutes),
                MinRefreshInterval = TimeSpan.FromMinutes(minRefreshIntervalMinutes),
                MaxFileAge = TimeSpan.FromDays(60),
                BasePath = basePath,
                UseAvailableValuesWhileRefreshing = true,
                ValidationMode = ValidationMode.Basic3G
            });
var  _blacklistProvider = ItalianDrlBlacklistProvider.Create(_httpClient,
                    new ItalianDrlBlacklistProviderOptions
                    {
                        RefreshInterval = TimeSpan.FromMinutes(refreshIntervalMinutes),
                        MinRefreshInterval = TimeSpan.FromMinutes(minRefreshIntervalMinutes),
                        MaxFileAge = TimeSpan.FromDays(60),
                        BasePath = basePath,
                        UseAvailableValuesWhileRefreshing = true,
                    });
var reader = DgcReaderService.Create(_trustListProvider, _blacklistProvider, _rulesValidator);
var res = await reader.VerifyForItaly(qr, ValidationMode.Basic3G, throwOnError: false);
```

The validation fails with Status `NeedRulesVerification` and `RulesValidation` field set to `null`.  The signature is correctly detected as Valid and the `Dgc` Name and Birth is correctly parsed. The `StatusMessage` field is: `Riferimento a un oggetto non impostato su un'istanza di oggetto.`

This is the Stack Trace I get by setting the `throwOnError` parameter to `true`:

```
   at DgcReader.DgcReaderService.<Verify>d__5.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at DgcReader.ItalianValidationResultsExtensions.<VerifyForItaly>d__4.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
```

I have already tried to re-download the updated rules, but with no luck. I am on version 2.5.0.
Am I doing something wrong? 
Please let me know if you need more details

**Comment by DevTrevi** [comment]
Hi @mircoianese , thank you for reporting this, is it possible for you to send me in private the exemption certificate?

**Comment by mircoianese** [comment]
> Hi @mircoianese , thank you for reporting this, is it possible for you to send me in private the exemption certificate?

Hi @DevTrevi,

unfortunately at the moment we can't send you the certificate. We are asking the owner for permission.

Meanwhile I did some investigations by debugging the source code. 
The Exemption schema is implemented only inside the `ItalianDGC` class, which is related to the RuleValidators.Italy module, but not to the BlacklistProviders.Italy module.

The first issue I encountered is that the call of this method does not return any certificate (so no certificate identifier for the Blacklist check to be completed):
![image](https://user-images.githubusercontent.com/22716366/154079479-4874be09-da9a-4a5b-bcbe-54ba801aaf79.png)

Then I think there is also another issue in the Rules Validation, the crash happen on this line:
![image](https://user-images.githubusercontent.com/22716366/154080053-85919374-1149-4671-abd8-223a2eb527e7.png)

Inside this function:
![image](https://user-images.githubusercontent.com/22716366/154080147-90e87f0a-55ef-466f-901f-27a9cfc75e8b.png)

Some data is still missing:
![image](https://user-images.githubusercontent.com/22716366/154080353-ad8ddbe7-9373-4b63-9614-bd47961b02bb.png)

And the exception thrown is: `Operazione o metodo non implementato.`


Hope this helps you identify the issue.
Please let me know if something else is needed.

Thanks

**Comment by DevTrevi** [comment]
I think that I have already found the issue, I missed to implement some properties in the ExemptionEntry...
It is possible for you to download the latest commit in develop branch and test it with the exemption certificate? I think that this should fix the issue, but I can't test with a real exemption certificate

**Comment by DevTrevi** [comment]
Mhm you are right... I missed that, I think that at this point I will need to make the deserializer injectable too, in order to allow deserialization of custom schema like the italian one... I will work on it this afternoon, hopefully it will be available in few hours

**Comment by mircoianese** [comment]
> Mhm you are right... I missed that, I think that at this point I will need to make the deserializer injectable too, in order to allow deserialization of custom schema like the italian one... I will work on it this afternoon, hopefully it will be available in few hours

Thank you a lot!

**Comment by DevTrevi** [comment]
Hi @mircoianese , can you try again with the latest version in develop? You can also download it from nuget, version 2.5.1-rc1

**Comment by mircoianese** [comment]
> Hi @mircoianese , can you try again with the latest version in develop? You can also download it from nuget, version 2.5.1-rc1

Hi @DevTrevi,

just tested it through nuget, it seems to be working!
I will wait for the 2.5.1 final release (which I see has also the School Mode removed), do you plan to release it today?
Thanks again!

**Comment by DevTrevi** [comment]
Hi @mircoianese , I have just released version 2.5.1
If you need to access Exemption details, you can try to cast the Dgc property of the result to `ItaliandDGC` with `.AsItalianDgc()` method
`ItalianDGC` inherits from `EuDGC`, and the certificate will be deserialized into this type when the issuer of the certificate is "IT"
Thank you for your help :)
